packages/framework/src/client.ts

```diff
@@ -95,21 +97,49 @@ export class Client {
     return builtConfiguration;
   }
 
+  /**
+   * Adds workflows to the client.
+   *
+   * A locking mechanism is used to ensure that duplicate workflows are not added.
+   *
+   * @param workflows - The workflows to add.
+   */
   public async addWorkflows(workflows: Array<Workflow>): Promise<void> {
     for (const workflow of workflows) {
-      if (this.discoveredWorkflows.some((existing) => existing.workflowId === workflow.id)) {
-        return;
-      } else {
-        const definition = await workflow.discover();
-        prettyPrintDiscovery(definition);
-        this.discoveredWorkflows.push(definition);
+      if (this.discoveredWorkflows.has(workflow.id)) {
+        continue;
+      }
+
+      if (this.addingWorkflows.has(workflow.id)) {
```

Comment:
Since the workflow promises Set is using the workflowId, do we need the addingWorkflows set? Can we use the map of promises as an indication of a workflow discovery already being in progress?

Reply:
Yes, you are right! I have removed the simple string map and we lean on the promises only now.